### PR TITLE
Integration with Azure Cognitive Search

### PR DESCRIPTION
**Pull request by UmerHA**
# Integration with Azure Cognitive Search

Allows to use Azure Cognitive Search as retriever ( & thus solves enhancement issue #3317).

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:
@hwchase17 @vowelparrot

### CODE REVIEW

File: langchain/retrievers/azure_cognitive_search.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate that api key exists in environment."""
+        api_key = get_from_dict_or_env(
+            values, "azure_cognitive_search_api_key", "AZURE_COGNITIVE_SEARCH_API_KEY"
+        )
+        values["azure_cognitive_search_api_key"] = api_key
+        return values
+
+
+    def _build_search_url(self, api_version: str = "2020-06-30") -> str:
```
Comment (dev2049) [inline langchain/retrievers/azure_cognitive_search.py:35]: should api_version be an instance variable? not sure how else someone would configure it

File: langchain/retrievers/azure_cognitive_search.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                    return await response.json()
+        else:
+            async with self.aiosession.get(search_url, headers=headers) as response:
+                return await response.json()
+
+    def get_relevant_documents(self, query: str) -> List[Document]:
+        search_results = self._search(query)
+
+        return [
+            Document(page_content=result["content"], metadata=result)
```
Comment (dev2049) [inline langchain/retrievers/azure_cognitive_search.py:69]: should we pop "content" from metadata

File: tests/integration_tests/retrievers/test_azure_cognitive_search.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from typing import List
+
+from langchain.schema import Document
+from langchain.retrievers.azure_cognitive_search import AzureCognitiveSearchRetriever
+
+
+def test_azure_cognitive_search_call() -> None:
+    """Test valid call to Azure Cognitive Search."""
+
+    assert os.environ.get("TEST_AZURE_COGNITIVE_SEARCH_SERVICE") is not None
```
Comment (dev2049) [inline tests/integration_tests/retrievers/test_azure_cognitive_search.py:13]: not very familiar with ACS but if it's not uncommon to store search_service and search_index as env vars should we make it possible to retriever those params from env vars (the way we do with api keys?)

**Review by dev2049 — COMMENTED** [review]
this is awesome @UmerHA! couple of small comments

File: langchain/retrievers/azure_cognitive_search.py (hunk 1 above)
Comment (UmerHA) [reply langchain/retrievers/azure_cognitive_search.py]: I agree - done ✅

File: langchain/retrievers/azure_cognitive_search.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import requests
+from pydantic import BaseModel, Extra, root_validator
+
+from langchain.schema import BaseRetriever, Document
+from langchain.utils import get_from_dict_or_env
+
+
+class AzureCognitiveSearchRetriever(BaseRetriever, BaseModel):
+    """Wrapper around Azure Cognitive Search."""
+    service_name: str  # name of Azure Cognitive Search service
```
[the remark this answers is not in the document]
Comment (UmerHA) [reply langchain/retrievers/azure_cognitive_search.py]: Done ✅

File: langchain/retrievers/azure_cognitive_search.py (hunk 2 above)
Comment (UmerHA) [reply langchain/retrievers/azure_cognitive_search.py]: I agree - done ✅

File: tests/integration_tests/retrievers/test_azure_cognitive_search.py (hunk 3 above)
Comment (UmerHA) [reply tests/integration_tests/retrievers/test_azure_cognitive_search.py]: I agree - done ✅

**Comment by dev2049** [comment]
this is huge, thanks @UmerHA!

**Pull request merged**